File: nanopdf-js/examples/02-text-extraction.ts
/**
 * Example 2: Text Extraction
 *
 * This example demonstrates how to:
 * - Extract text from all pages
 * - Search for specific text
 * - Extract text with layout information
 * - Export text to file
 */

import { Document } from 'nanopdf';
import { resolve } from 'path';
import { writeFileSync } from 'fs';

function extractAllText() {
  const pdfPath = resolve(__dirname, '../../test-pdfs/simple/multi-page.pdf');
  const doc = Document.open(pdfPath);

  console.log('=== Extracting Text from All Pages ===\n');

  let allText = '';

  for (let i = 0; i < doc.pageCount; i++) {
    const page = doc.loadPage(i);
    const text = page.extractText();

    console.log(`Page ${i + 1}:`);
    console.log('-'.repeat(50));
    console.log(text);
    console.log('\n');

    allText += `\n=== Page ${i + 1} ===\n${text}\n`;

    page.drop();
  }

  // Save to file
  const outputPath = resolve(__dirname, 'extracted-text.txt');
  writeFileSync(outputPath, allText);
  console.log(`✅ Saved to: ${outputPath}`);

  doc.close();
}

function searchText() {
  const pdfPath = resolve(__dirname, '../../test-pdfs/simple/multi-page.pdf');
  const doc = Document.open(pdfPath);

  console.log('\n=== Searching for Text ===\n');

  const searchTerm = 'Page';

  for (let i = 0; i < doc.pageCount; i++) {
    const page = doc.loadPage(i);
    const hits = page.searchText(searchTerm);

    if (hits.length > 0) {
      console.log(`Page ${i + 1}: Found ${hits.length} occurrence(s)`);

      hits.forEach((rect, index) => {
        console.log(`  Hit ${index + 1}: [${rect.x0.toFixed(1)}, ${rect.y0.toFixed(1)}, ${rect.x1.toFixed(1)}, ${rect.y1.toFixed(1)}]`);
      });
    }

    page.drop();
  }

  doc.close();
  console.log('\n✅ Done!');
}

function extractTextBlocks() {
  const pdfPath = resolve(__dirname, '../../test-pdfs/medium/with-metadata.pdf');
  const doc = Document.open(pdfPath);

  console.log('\n=== Extracting Text Blocks with Layout ===\n');

  const page = doc.loadPage(0);
  const blocks = page.extractTextBlocks();

  console.log(`Found ${blocks.length} text block(s)\n`);

  blocks.forEach((block, index) => {
    console.log(`Block ${index + 1}:`);
    console.log(`  Bounds: [${block.bbox[0].toFixed(1)}, ${block.bbox[1].toFixed(1)}, ${block.bbox[2].toFixed(1)}, ${block.bbox[3].toFixed(1)}]`);
    console.log(`  Lines: ${block.lines.length}`);

    block.lines.forEach((line, lineIdx) => {
      const text = line.spans.map(s => s.text).join('');
      console.log(`    Line ${lineIdx + 1}: "${text}"`);
    });

    console.log('');
  });

  page.drop();
  doc.close();

  console.log('✅ Done!');
}

// Run examples
if (require.main === module) {
  try {
    extractAllText();
    searchText();
    extractTextBlocks();
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

export { extractAllText, searchText, extractTextBlocks };

File: go-nanopdf/examples/02_text_extraction.go
// Example 2: Text Extraction
//
// This example demonstrates how to:
//   - Extract text from all pages
//   - Search for specific text
//   - Save extracted text to file

//go:build example

package main

import (
	"fmt"
	"log"
	"os"
	"path/filepath"

	nanopdf "github.com/lexmata/nanopdf/go-nanopdf"
)

func main() {
	ctx := nanopdf.NewContext()
	if ctx == nil {
		log.Fatal("Failed to create context")
	}
	defer ctx.Drop()

	pdfPath := filepath.Join("..", "..", "test-pdfs", "simple", "multi-page.pdf")
	doc, err := nanopdf.OpenDocument(ctx, pdfPath)
	if err != nil {
		log.Fatalf("Failed to open PDF: %v", err)
	}
	defer doc.Drop()

	fmt.Println("=== Extracting Text from All Pages ===\n")

	pageCount, err := doc.PageCount()
	if err != nil {
		log.Fatal(err)
	}

	var allText string

	for i := int32(0); i < pageCount; i++ {
		page, err := doc.LoadPage(i)
		if err != nil {
			log.Printf("Error loading page %d: %v", i, err)
			continue
		}

		text, err := page.ExtractText()
		if err != nil {
			log.Printf("Error extracting text from page %d: %v", i, err)
			page.Drop()
			continue
		}

		fmt.Printf("Page %d:\n", i+1)
		fmt.Println("--------------------------------------------------")
		fmt.Println(text)
		fmt.Println()

		allText += fmt.Sprintf("\n=== Page %d ===\n%s\n", i+1, text)

		page.Drop()
	}

	// Save to file
	outputPath := "extracted-text.txt"
	err = os.WriteFile(outputPath, []byte(allText), 0644)
	if err != nil {
		log.Printf("Warning: Could not save to file: %v", err)
	} else {
		fmt.Printf("✅ Saved to: %s\n", outputPath)
	}

	// Search for text
	fmt.Println("\n=== Searching for Text ===\n")

	searchTerm := "Page"
	totalHits := 0

	for i := int32(0); i < pageCount; i++ {
		page, err := doc.LoadPage(i)
		if err != nil {
			continue
		}

		results, err := page.SearchText(searchTerm)
		if err != nil {
			page.Drop()
			continue
		}

		if len(results) > 0 {
			fmt.Printf("Page %d: Found %d occurrence(s)\n", i+1, len(results))

			for idx, rect := range results {
				if idx < 3 { // Show first 3 hits
					fmt.Printf("  Hit %d: [%.1f, %.1f, %.1f, %.1f]\n",
						idx+1, rect.X0, rect.Y0, rect.X1, rect.Y1)
				}
			}
			totalHits += len(results)
		}

		page.Drop()
	}

	fmt.Printf("\nTotal: %d occurrences found\n", totalHits)
	fmt.Println("\n✅ Done!")
}

